File: sdk/javascript/src/impl/CipherHelpers.js
import { AesCbcCipher, AesGcmCipher } from '../Cipher.js';
import crypto from 'crypto';

const GCM_IV_SIZE = 12;
const CBC_IV_SIZE = 16;
const SALT_SIZE = 32;

export const concatArrays = (...arrays) => {
	const totalLength = arrays.map(buffer => buffer.length).reduce((accumulator, currentValue) => accumulator + currentValue);
	const result = new Uint8Array(totalLength);
	let targetOffset = 0;
	arrays.forEach(buffer => {
		result.set(buffer, targetOffset);
		targetOffset += buffer.length;
	});
	return result;
};

const decode = (tagSize, ivSize, encodedMessage) => ({
	tag: encodedMessage.subarray(0, tagSize),
	initializationVector: encodedMessage.subarray(tagSize, tagSize + ivSize),
	encodedMessageData: encodedMessage.subarray(tagSize + ivSize)
});

export const decodeAesGcm = (deriveSharedKey, keyPair, recipientPublicKey, encodedMessage) => {
	const { tag, initializationVector, encodedMessageData } = decode(AesGcmCipher.TAG_SIZE, GCM_IV_SIZE, encodedMessage);

	const sharedKey = deriveSharedKey(keyPair, recipientPublicKey);
	const cipher = new AesGcmCipher(sharedKey);

	return new Uint8Array(cipher.decrypt(concatArrays(encodedMessageData, tag), initializationVector));
};

export const decodeAesCbc = (deriveSharedKey, keyPair, recipientPublicKey, encodedMessage) => {
	const { tag, initializationVector, encodedMessageData } = decode(SALT_SIZE, CBC_IV_SIZE, encodedMessage);

	const sharedKey = deriveSharedKey(keyPair, recipientPublicKey, tag);
	const cipher = new AesCbcCipher(sharedKey);

	return new Uint8Array(cipher.decrypt(encodedMessageData, initializationVector));
};

export const encodeAesGcm = (deriveSharedKey, keyPair, recipientPublicKey, message) => {
	const sharedKey = deriveSharedKey(keyPair, recipientPublicKey);
	const cipher = new AesGcmCipher(sharedKey);

	const initializationVector = new Uint8Array(crypto.randomBytes(GCM_IV_SIZE));
	const cipherText = cipher.encrypt(message, initializationVector);
	const tagStartOffset = cipherText.length - AesGcmCipher.TAG_SIZE;
	const tag = cipherText.subarray(tagStartOffset);

	return { tag, initializationVector, cipherText: cipherText.subarray(0, tagStartOffset) };
};

export const encodeAesCbc = (deriveSharedKey, keyPair, recipientPublicKey, message) => {
	const salt = new Uint8Array(crypto.randomBytes(SALT_SIZE));
	const sharedKey = deriveSharedKey(keyPair, recipientPublicKey, salt);
	const cipher = new AesCbcCipher(sharedKey);

	const initializationVector = new Uint8Array(crypto.randomBytes(CBC_IV_SIZE));
	const cipherText = cipher.encrypt(message, initializationVector);

	return { salt, initializationVector, cipherText };
};

File: sdk/cpp/symbolchain/impl/CipherHelpers.py
import secrets

from symbolchain.Cipher import AesCbcCipher, AesGcmCipher

GCM_IV_SIZE = 12
CBC_IV_SIZE = 16
SALT_SIZE = 32


def _decode(tag_or_salt_size, iv_size, encoded_message):
	tag_or_salt = encoded_message[:tag_or_salt_size]
	initialization_vector = encoded_message[tag_or_salt_size:tag_or_salt_size + iv_size]
	encoded_message_data = encoded_message[tag_or_salt_size + iv_size:]

	return tag_or_salt, initialization_vector, encoded_message_data


def decode_aes_gcm(shared_key_class, key_pair, recipient_public_key, encoded_message):
	tag, initialization_vector, encoded_message_data = _decode(AesGcmCipher.TAG_SIZE, GCM_IV_SIZE, encoded_message)

	shared_key = shared_key_class.derive_shared_key(key_pair, recipient_public_key)
	cipher = AesGcmCipher(shared_key)

	return cipher.decrypt(encoded_message_data + tag, initialization_vector)


def decode_aes_cbc(shared_key_class, key_pair, recipient_public_key, encoded_message):
	salt, initialization_vector, encoded_message_data = _decode(SALT_SIZE, CBC_IV_SIZE, encoded_message)

	shared_key = shared_key_class.derive_shared_key_deprecated(key_pair, recipient_public_key, salt)
	cipher = AesCbcCipher(shared_key)

	return cipher.decrypt(encoded_message_data, initialization_vector)


def encode_aes_gcm(shared_key_class, key_pair, recipient_public_key, message):
	shared_key = shared_key_class.derive_shared_key(key_pair, recipient_public_key)
	cipher = AesGcmCipher(shared_key)

	initialization_vector = secrets.token_bytes(GCM_IV_SIZE)
	cipher_text = cipher.encrypt(message, initialization_vector)

	tag_start_offset = len(cipher_text) - AesGcmCipher.TAG_SIZE
	tag = cipher_text[tag_start_offset:]

	return tag, initialization_vector, cipher_text[:tag_start_offset]


def encode_aes_cbc(shared_key_class, key_pair, recipient_public_key, message):
	salt = secrets.token_bytes(SALT_SIZE)
	shared_key = shared_key_class.derive_shared_key_deprecated(key_pair, recipient_public_key, salt)
	cipher = AesCbcCipher(shared_key)

	initialization_vector = secrets.token_bytes(CBC_IV_SIZE)
	cipher_text = cipher.encrypt(message, initialization_vector)

	return salt, initialization_vector, cipher_text
